File: android/app/src/main/java/com/kreativekoala/riddleverse/PinballDeflectorGenerator.kt
package com.kreativekoala.riddleverse

import kotlin.math.abs
import kotlin.random.Random

/**
 * Local generator for Pinball Deflector puzzles
 * Generates matrix with deflectors that change ball direction by 90 degrees
 * User must predict where ball will end up after deflections
 */
object PinballDeflectorGenerator {

    // Deflector orientation constants
    object DeflectorTypes {
        const val EMPTY = 0
        const val SLASH = 1      // / deflector
        const val BACKSLASH = 2  // \ deflector
    }

    // Direction data class
    data class Direction(
        val dr: Int,
        val dc: Int,
        val name: String
    )

    // Predefined directions
    val DIRECTIONS = mapOf(
        "UP" to Direction(-1, 0, "UP"),
        "DOWN" to Direction(1, 0, "DOWN"),
        "LEFT" to Direction(0, -1, "LEFT"),
        "RIGHT" to Direction(0, 1, "RIGHT")
    )

    data class DifficultyConfig(
        val matrixSize: Int,
        val deflectorCount: Int,
        val minActiveDeflectors: Int,
        val memoryTime: Int,
        val maxBounces: Int
    )

    data class BallPathStep(
        val position: List<Int>,
        val direction: Direction,
        val deflected: Boolean,
        val deflectorType: Int?
    )

    data class ActiveDeflector(
        val position: List<Int>,
        val type: Int,
        val oldDirection: Direction,
        val newDirection: Direction
    )

    private val difficultyConfig = mapOf(
        "easy" to DifficultyConfig(
            matrixSize = 5,
            deflectorCount = 3,
            minActiveDeflectors = 2,
            memoryTime = 4000,
            maxBounces = 8
        ),
        "medium" to DifficultyConfig(
            matrixSize = 6,
            deflectorCount = 4,
            minActiveDeflectors = 3,
            memoryTime = 5000,
            maxBounces = 10
        ),
        "hard" to DifficultyConfig(
            matrixSize = 7,
            deflectorCount = 5,
            minActiveDeflectors = 3,
            memoryTime = 6000,
            maxBounces = 12
        )
    )

    /**
     * Generate a complete Pinball Deflector puzzle
     */
    fun generatePuzzle(difficulty: String = "easy"): Pair<String, String> {
        val config = difficultyConfig[difficulty.lowercase()]
            ?: difficultyConfig["easy"]!!

        println("🎯 Generating $difficulty pinball deflector puzzle")

        // Generate puzzle with valid ball path
        val puzzleResult = generatePuzzleWithPath(config)

        if (!puzzleResult.success) {
            throw Exception(puzzleResult.error ?: "Failed to generate valid puzzle")
        }

        val matrix = puzzleResult.matrix!!
        val startPosition = puzzleResult.startPosition!!
        val startDirection = puzzleResult.startDirection!!
        val ballPath = puzzleResult.ballPath!!
        val endPosition = puzzleResult.endPosition!!
        val activeDeflectors = puzzleResult.activeDeflectors!!

        // Create instructions
        val instructions = mapOf(
            "title" to "Pinball Deflector Puzzle",
            "description" to "Predict where the ball will end up after hitting deflectors!",
            "steps" to listOf(
                "1. Memorize the deflector positions and orientations",
                "2. Deflectors will disappear after the timer",
                "3. A ball will appear at the start position",
                "4. Predict where the ball will end up after all deflections",
                "5. Tap the predicted end position"
            ),
            "tip" to "You have ${config.memoryTime / 1000} seconds to memorize the ${config.deflectorCount} deflectors"
        )

        // Build question JSON
        val questionData = mapOf(
            "matrix" to matrix,
            "matrixSize" to config.matrixSize,
            "startPosition" to startPosition,
            "startDirection" to mapOf(
                "dr" to startDirection.dr,
                "dc" to startDirection.dc,
                "name" to startDirection.name
            ),
            "memoryTime" to config.memoryTime,
            "instructions" to instructions,
            "difficulty" to difficulty
        )

        // Build answer JSON
        val answerData = mapOf(
            "endPosition" to endPosition,
            "ballPath" to ballPath.map { step ->
                mapOf(
                    "position" to step.position,
                    "direction" to mapOf(
                        "dr" to step.direction.dr,
                        "dc" to step.direction.dc,
                        "name" to step.direction.name
                    ),
                    "deflected" to step.deflected,
                    "deflectorType" to step.deflectorType
                )
            },
            "activeDeflectors" to activeDeflectors.map { deflector ->
                mapOf(
                    "position" to deflector.position,
                    "type" to deflector.type,
                    "oldDirection" to mapOf(
                        "dr" to deflector.oldDirection.dr,
                        "dc" to deflector.oldDirection.dc,
                        "name" to deflector.oldDirection.name
                    ),
                    "newDirection" to mapOf(
                        "dr" to deflector.newDirection.dr,
                        "dc" to deflector.newDirection.dc,
                        "name" to deflector.newDirection.name
                    )
                )
            },
            "scoring" to mapOf(
                "correctEndBonus" to 70,
                "pathAccuracyBonus" to 30,
                "maxPoints" to 100
            )
        )

        val questionJson = buildJsonString(questionData)
        val answerJson = buildJsonString(answerData)

        println("✅ Generated ${config.matrixSize}x${config.matrixSize} matrix with ${countDeflectors(matrix)} deflectors")
        println("🎯 Ball path: ${ballPath.size} steps, ${ballPath.count { it.deflected }} bounces")
        println("📍 Start: [${startPosition[0]},${startPosition[1]}] ${startDirection.name}, End: [${endPosition[0]},${endPosition[1]}]")

        return Pair(questionJson, answerJson)
    }

    /**
     * Generate puzzle ensuring ball hits minimum deflectors
     */
    private fun generatePuzzleWithPath(config: DifficultyConfig): PuzzleResult {
        var attempts = 0
        val maxAttempts = 50

        while (attempts < maxAttempts) {
            attempts++

            // Create matrix with deflectors
            val matrix = createEmptyMatrix(config.matrixSize)
            val startResult = selectRandomStartPosition(config.matrixSize)

            if (!startResult.success) {
                continue
            }

            val startPosition = startResult.position!!
            val startDirection = startResult.direction!!

            // Place deflectors strategically
            val deflectorResult = placeStrategicDeflectors(matrix, startPosition, startDirection, config)

            if (!deflectorResult.success) {
                continue
            }

            // Simulate ball path
            val pathResult = simulateBallPath(matrix, startPosition, startDirection, config.maxBounces)

            if (!pathResult.success) {
                continue
            }

            val ballPath = pathResult.path!!
            val endPosition = pathResult.endPosition!!
            val activeDeflectors = pathResult.activeDeflectors!!

            // Check if minimum deflectors were hit
            if (activeDeflectors.size >= config.minActiveDeflectors) {
                println("✅ Valid puzzle generated on attempt $attempts")
                return PuzzleResult(
                    success = true,
                    matrix = matrix,
                    startPosition = startPosition,
                    startDirection = startDirection,
                    ballPath = ballPath,
                    endPosition = endPosition,
                    activeDeflectors = activeDeflectors
                )
            }
        }

        return PuzzleResult(
            success = false,
            error = "Could not generate valid puzzle after $maxAttempts attempts"
        )
    }

    /**
     * Create empty matrix
     */
    private fun createEmptyMatrix(size: Int): List<MutableList<Int>> {
        return List(size) { MutableList(size) { DeflectorTypes.EMPTY } }
    }

    /**
     * Select random start position on matrix edge
     */
    private fun selectRandomStartPosition(matrixSize: Int): StartPositionResult {
        val edges = mutableListOf<Pair<List<Int>, Direction>>()

        // Top edge (going down)
        for (i in 0 until matrixSize) {
            edges.add(Pair(listOf(0, i), DIRECTIONS["DOWN"]!!))
        }

        // Bottom edge (going up)
        for (i in 0 until matrixSize) {
            edges.add(Pair(listOf(matrixSize - 1, i), DIRECTIONS["UP"]!!))
        }

        // Left edge (going right)
        for (i in 0 until matrixSize) {
            edges.add(Pair(listOf(i, 0), DIRECTIONS["RIGHT"]!!))
        }

        // Right edge (going left)
        for (i in 0 until matrixSize) {
            edges.add(Pair(listOf(i, matrixSize - 1), DIRECTIONS["LEFT"]!!))
        }

        val randomEdge = edges.random()
        return StartPositionResult(
            success = true,
            position = randomEdge.first,
            direction = randomEdge.second
        )
    }

    /**
     * Place deflectors strategically in ball's path
     */
    private fun placeStrategicDeflectors(
        matrix: List<MutableList<Int>>,
        startPosition: List<Int>,
        startDirection: Direction,
        config: DifficultyConfig
    ): PlacementResult {
        val matrixSize = config.matrixSize
        val deflectorCount = config.deflectorCount
        val minActiveDeflectors = config.minActiveDeflectors

        var currentPos = startPosition.toMutableList()
        var currentDir = startDirection.copy()
        var deflatorsPlaced = 0

        // Simulate initial path to place first deflector
        while (deflatorsPlaced < minOf(deflectorCount, minActiveDeflectors)) {
            // Move in current direction to find placement spot
            val nextPos = listOf(
                currentPos[0] + currentDir.dr,
                currentPos[1] + currentDir.dc
            )

            // Check bounds
            if (nextPos[0] < 0 || nextPos[0] >= matrixSize ||
                nextPos[1] < 0 || nextPos[1] >= matrixSize) {
                break
            }

            // Skip if too close to start (first 1-2 cells)
            val distanceFromStart = abs(nextPos[0] - startPosition[0]) + abs(nextPos[1] - startPosition[1])
            if (distanceFromStart <= 1) {
                currentPos = nextPos.toMutableList()
                continue
            }

            // Place deflector with some probability
            if (Random.nextDouble() < 0.4 && matrix[nextPos[0]][nextPos[1]] == DeflectorTypes.EMPTY) {
                // Randomly choose deflector type
                val deflectorType = if (Random.nextBoolean()) DeflectorTypes.SLASH else DeflectorTypes.BACKSLASH
                matrix[nextPos[0]][nextPos[1]] = deflectorType
                deflatorsPlaced++

                // Update direction based on deflector
                currentDir = getNewDirection(currentDir, deflectorType)
                println("📍 Placed deflector $deflectorType at [${nextPos[0]},${nextPos[1]}], new direction: ${currentDir.name}")
            }

            currentPos = nextPos.toMutableList()
        }

        // Place remaining deflectors randomly
        while (deflatorsPlaced < deflectorCount) {
            val row = Random.nextInt(matrixSize)
            val col = Random.nextInt(matrixSize)

            if (matrix[row][col] == DeflectorTypes.EMPTY) {
                val deflectorType = if (Random.nextBoolean()) DeflectorTypes.SLASH else DeflectorTypes.BACKSLASH
                matrix[row][col] = deflectorType
                deflatorsPlaced++
            }
        }

        return PlacementResult(success = true)
    }

    /**
     * Simulate ball path through matrix
     */
    private fun simulateBallPath(
        matrix: List<List<Int>>,
        startPosition: List<Int>,
        startDirection: Direction,
        maxBounces: Int
    ): PathResult {
        val matrixSize = matrix.size
        val path = mutableListOf<BallPathStep>()
        val activeDeflectors = mutableListOf<ActiveDeflector>()

        var currentPos = startPosition.toMutableList()
        var currentDir = startDirection.copy()
        var bounces = 0

        // Add start position
        path.add(BallPathStep(
            position = currentPos.toList(),
            direction = currentDir.copy(),
            deflected = false,
            deflectorType = null
        ))

        while (bounces < maxBounces) {
            // Calculate next position
            val nextPos = listOf(
                currentPos[0] + currentDir.dr,
                currentPos[1] + currentDir.dc
            )

            // Check if out of bounds
            if (nextPos[0] < 0 || nextPos[0] >= matrixSize ||
                nextPos[1] < 0 || nextPos[1] >= matrixSize) {
                break
            }

            currentPos = nextPos.toMutableList()
            val cellValue = matrix[currentPos[0]][currentPos[1]]

            // Check for deflector
            if (cellValue == DeflectorTypes.SLASH || cellValue == DeflectorTypes.BACKSLASH) {
                // Ball hits deflector
                val newDirection = getNewDirection(currentDir, cellValue)

                path.add(BallPathStep(
                    position = currentPos.toList(),
                    direction = currentDir.copy(),
                    deflected = true,
                    deflectorType = cellValue
                ))

                activeDeflectors.add(ActiveDeflector(
                    position = currentPos.toList(),
                    type = cellValue,
                    oldDirection = currentDir.copy(),
                    newDirection = newDirection.copy()
                ))

                currentDir = newDirection
                bounces++
            } else {
                // Empty cell
                path.add(BallPathStep(
                    position = currentPos.toList(),
                    direction = currentDir.copy(),
                    deflected = false,
                    deflectorType = null
                ))
            }
        }

        val endPosition = path.last().position

        return PathResult(
            success = true,
            path = path,
            endPosition = endPosition,
            activeDeflectors = activeDeflectors
        )
    }

    /**
     * Calculate new direction after hitting deflector
     */
    private fun getNewDirection(currentDirection: Direction, deflectorType: Int): Direction {
        val dr = currentDirection.dr
        val dc = currentDirection.dc

        return when (deflectorType) {
            DeflectorTypes.SLASH -> {
                // Slash deflector: (dr, dc) -> (-dc, -dr)
                getDirectionFromComponents(-dc, -dr)
            }
            DeflectorTypes.BACKSLASH -> {
                // Backslash deflector: (dr, dc) -> (dc, dr)
                getDirectionFromComponents(dc, dr)
            }
            else -> currentDirection // No change
        }
    }

    /**
     * Get direction object from components
     */
    private fun getDirectionFromComponents(dr: Int, dc: Int): Direction {
        return when {
            dr == -1 && dc == 0 -> DIRECTIONS["UP"]!!
            dr == 1 && dc == 0 -> DIRECTIONS["DOWN"]!!
            dr == 0 && dc == -1 -> DIRECTIONS["LEFT"]!!
            dr == 0 && dc == 1 -> DIRECTIONS["RIGHT"]!!
            else -> Direction(dr, dc, "CUSTOM_${dr}_${dc}")
        }
    }

    /**
     * Count deflectors in matrix
     */
    private fun countDeflectors(matrix: List<List<Int>>): Int {
        return matrix.flatten().count { cell ->
            cell == DeflectorTypes.SLASH || cell == DeflectorTypes.BACKSLASH
        }
    }

    /**
     * Build JSON string from nested data structures
     */
    private fun buildJsonString(data: Any?): String {
        return when (data) {
            is Map<*, *> -> {
                val entries = data.entries.joinToString(",") { (key, value) ->
                    "\"$key\":${buildJsonString(value)}"
                }
                "{$entries}"
            }
            is List<*> -> {
                val elements = data.joinToString(",") { buildJsonString(it) }
                "[$elements]"
            }
            is String -> "\"${data.replace("\"", "\\\"")}\""
            is Number -> data.toString()
            is Boolean -> data.toString()
            null -> "null"
            else -> "\"$data\""
        }
    }

    // Helper data classes
    data class PuzzleResult(
        val success: Boolean,
        val matrix: List<MutableList<Int>>? = null,
        val startPosition: List<Int>? = null,
        val startDirection: Direction? = null,
        val ballPath: List<BallPathStep>? = null,
        val endPosition: List<Int>? = null,
        val activeDeflectors: List<ActiveDeflector>? = null,
        val error: String? = null
    )

    data class StartPositionResult(
        val success: Boolean,
        val position: List<Int>? = null,
        val direction: Direction? = null
    )

    data class PlacementResult(
        val success: Boolean
    )

    data class PathResult(
        val success: Boolean,
        val path: List<BallPathStep>? = null,
        val endPosition: List<Int>? = null,
        val activeDeflectors: List<ActiveDeflector>? = null
    )
}

File: ios/PuzzleForge/PinballDeflectorGenerator.swift
//
//  PinballDeflectorGenerator.swift
//  PuzzleForge
//
//  Created by Assistant on 1/15/25.
//

import Foundation

/**
 * Local generator for Pinball Deflector puzzles
 * Generates matrix with deflectors that change ball direction by 90 degrees
 * User must predict where ball will end up after deflections
 */
class PinballDeflectorGenerator {
    
    // Deflector orientation constants
    struct DeflectorTypes {
        static let empty = 0
        static let slash = 1      // / deflector
        static let backslash = 2  // \ deflector
    }
    
    // Direction data structure
    struct Direction {
        let dr: Int
        let dc: Int
        let name: String
        
        func copy() -> Direction {
            return Direction(dr: dr, dc: dc, name: name)
        }
    }
    
    // Predefined directions
    static let directions: [String: Direction] = [
        "UP": Direction(dr: -1, dc: 0, name: "UP"),
        "DOWN": Direction(dr: 1, dc: 0, name: "DOWN"),
        "LEFT": Direction(dr: 0, dc: -1, name: "LEFT"),
        "RIGHT": Direction(dr: 0, dc: 1, name: "RIGHT")
    ]
    
    struct DifficultyConfig {
        let matrixSize: Int
        let deflectorCount: Int
        let minActiveDeflectors: Int
        let memoryTime: Int
        let maxBounces: Int
    }
    
    struct BallPathStep {
        let position: [Int]
        let direction: Direction
        let deflected: Bool
        let deflectorType: Int?
    }
    
    struct ActiveDeflector {
        let position: [Int]
        let type: Int
        let oldDirection: Direction
        let newDirection: Direction
    }
    
    private static let difficultyConfig: [String: DifficultyConfig] = [
        "easy": DifficultyConfig(
            matrixSize: 5,
            deflectorCount: 3,
            minActiveDeflectors: 2,
            memoryTime: 4000,
            maxBounces: 8
        ),
        "medium": DifficultyConfig(
            matrixSize: 6,
            deflectorCount: 4,
            minActiveDeflectors: 3,
            memoryTime: 5000,
            maxBounces: 10
        ),
        "hard": DifficultyConfig(
            matrixSize: 7,
            deflectorCount: 5,
            minActiveDeflectors: 3,
            memoryTime: 6000,
            maxBounces: 12
        )
    ]
    
    /**
     * Generate a complete Pinball Deflector puzzle
     */
    static func generatePuzzle(difficulty: String = "easy") -> (String, String) {
        let config = difficultyConfig[difficulty.lowercased()] ?? difficultyConfig["easy"]!
        
        print("🎯 Generating \(difficulty) pinball deflector puzzle")
        
        // Generate puzzle with valid ball path
        let puzzleResult = generatePuzzleWithPath(config: config)
        
        guard puzzleResult.success,
              let matrix = puzzleResult.matrix,
              let startPosition = puzzleResult.startPosition,
              let startDirection = puzzleResult.startDirection,
              let ballPath = puzzleResult.ballPath,
              let endPosition = puzzleResult.endPosition,
              let activeDeflectors = puzzleResult.activeDeflectors else {
            fatalError(puzzleResult.error ?? "Failed to generate valid puzzle")
        }
        
        // Create instructions
        let instructions: [String: Any] = [
            "title": "Pinball Deflector Puzzle",
            "description": "Predict where the ball will end up after hitting deflectors!",
            "steps": [
                "1. Memorize the deflector positions and orientations",
                "2. Deflectors will disappear after the timer",
                "3. A ball will appear at the start position",
                "4. Predict where the ball will end up after all deflections",
                "5. Tap the predicted end position"
            ],
            "tip": "You have \(config.memoryTime / 1000) seconds to memorize the \(config.deflectorCount) deflectors"
        ]
        
        // Build question JSON
        let questionData: [String: Any] = [
            "matrix": matrix,
            "matrixSize": config.matrixSize,
            "startPosition": startPosition,
            "startDirection": [
                "dr": startDirection.dr,
                "dc": startDirection.dc,
                "name": startDirection.name
            ],
            "memoryTime": config.memoryTime,
            "instructions": instructions,
            "difficulty": difficulty
        ]
        
        // Build answer JSON
        let answerData: [String: Any] = [
            "endPosition": endPosition,
            "ballPath": ballPath.map { step in
                [
                    "position": step.position,
                    "direction": [
                        "dr": step.direction.dr,
                        "dc": step.direction.dc,
                        "name": step.direction.name
                    ],
                    "deflected": step.deflected,
                    "deflectorType": step.deflectorType as Any
                ]
            },
            "activeDeflectors": activeDeflectors.map { deflector in
                [
                    "position": deflector.position,
                    "type": deflector.type,
                    "oldDirection": [
                        "dr": deflector.oldDirection.dr,
                        "dc": deflector.oldDirection.dc,
                        "name": deflector.oldDirection.name
                    ],
                    "newDirection": [
                        "dr": deflector.newDirection.dr,
                        "dc": deflector.newDirection.dc,
                        "name": deflector.newDirection.name
                    ]
                ]
            },
            "scoring": [
                "correctEndBonus": 70,
                "pathAccuracyBonus": 30,
                "maxPoints": 100
            ]
        ]
        
        let questionJson = buildJsonString(data: questionData)
        let answerJson = buildJsonString(data: answerData)
        
        print("✅ Generated \(config.matrixSize)x\(config.matrixSize) matrix with \(countDeflectors(matrix: matrix)) deflectors")
        print("🎯 Ball path: \(ballPath.count) steps, \(ballPath.filter { $0.deflected }.count) bounces")
        print("📍 Start: [\(startPosition[0]),\(startPosition[1])] \(startDirection.name), End: [\(endPosition[0]),\(endPosition[1])]")
        
        return (questionJson, answerJson)
    }
    
    /**
     * Generate puzzle ensuring ball hits minimum deflectors
     */
    private static func generatePuzzleWithPath(config: DifficultyConfig) -> PuzzleResult {
        var attempts = 0
        let maxAttempts = 50
        
        while attempts < maxAttempts {
            attempts += 1
            
            // Create matrix with deflectors
            var matrix = createEmptyMatrix(size: config.matrixSize)
            let startResult = selectRandomStartPosition(matrixSize: config.matrixSize)
            
            guard startResult.success,
                  let startPosition = startResult.position,
                  let startDirection = startResult.direction else {
                continue
            }
            
            // Place deflectors strategically
            let deflectorResult = placeStrategicDeflectors(
                matrix: &matrix,
                startPosition: startPosition,
                startDirection: startDirection,
                config: config
            )
            
            guard deflectorResult.success else {
                continue
            }
            
            // Simulate ball path
            let pathResult = simulateBallPath(
                matrix: matrix,
                startPosition: startPosition,
                startDirection: startDirection,
                maxBounces: config.maxBounces
            )
            
            guard pathResult.success,
                  let ballPath = pathResult.path,
                  let endPosition = pathResult.endPosition,
                  let activeDeflectors = pathResult.activeDeflectors else {
                continue
            }
            
            // Check if minimum deflectors were hit
            if activeDeflectors.count >= config.minActiveDeflectors {
                print("✅ Valid puzzle generated on attempt \(attempts)")
                return PuzzleResult(
                    success: true,
                    matrix: matrix,
                    startPosition: startPosition,
                    startDirection: startDirection,
                    ballPath: ballPath,
                    endPosition: endPosition,
                    activeDeflectors: activeDeflectors
                )
            }
        }
        
        return PuzzleResult(
            success: false,
            error: "Could not generate valid puzzle after \(maxAttempts) attempts"
        )
    }
    
    /**
     * Create empty matrix
     */
    private static func createEmptyMatrix(size: Int) -> [[Int]] {
        return Array(repeating: Array(repeating: DeflectorTypes.empty, count: size), count: size)
    }
    
    /**
     * Select random start position on matrix edge
     */
    private static func selectRandomStartPosition(matrixSize: Int) -> StartPositionResult {
        var edges: [([Int], Direction)] = []
        
        // Top edge (going down)
        for i in 0..<matrixSize {
            edges.append(([0, i], directions["DOWN"]!))
        }
        
        // Bottom edge (going up)
        for i in 0..<matrixSize {
            edges.append(([matrixSize - 1, i], directions["UP"]!))
        }
        
        // Left edge (going right)
        for i in 0..<matrixSize {
            edges.append(([i, 0], directions["RIGHT"]!))
        }
        
        // Right edge (going left)
        for i in 0..<matrixSize {
            edges.append(([i, matrixSize - 1], directions["LEFT"]!))
        }
        
        let randomEdge = edges.randomElement()!
        return StartPositionResult(
            success: true,
            position: randomEdge.0,
            direction: randomEdge.1
        )
    }
    
    /**
     * Place deflectors strategically in ball's path
     */
    private static func placeStrategicDeflectors(
        matrix: inout [[Int]],
        startPosition: [Int],
        startDirection: Direction,
        config: DifficultyConfig
    ) -> PlacementResult {
        let matrixSize = config.matrixSize
        let deflectorCount = config.deflectorCount
        let minActiveDeflectors = config.minActiveDeflectors
        
        var currentPos = startPosition
        var currentDir = startDirection.copy()
        var deflatorsPlaced = 0
        
        // Simulate initial path to place first deflector
        while deflatorsPlaced < min(deflectorCount, minActiveDeflectors) {
            // Move in current direction to find placement spot
            let nextPos = [
                currentPos[0] + currentDir.dr,
                currentPos[1] + currentDir.dc
            ]
            
            // Check bounds
            if nextPos[0] < 0 || nextPos[0] >= matrixSize ||
               nextPos[1] < 0 || nextPos[1] >= matrixSize {
                break
            }
            
            // Skip if too close to start (first 1-2 cells)
            let distanceFromStart = abs(nextPos[0] - startPosition[0]) + abs(nextPos[1] - startPosition[1])
            if distanceFromStart <= 1 {
                currentPos = nextPos
                continue
            }
            
            // Place deflector with some probability
            if Double.random(in: 0...1) < 0.4 && matrix[nextPos[0]][nextPos[1]] == DeflectorTypes.empty {
                // Randomly choose deflector type
                let deflectorType = Bool.random() ? DeflectorTypes.slash : DeflectorTypes.backslash
                matrix[nextPos[0]][nextPos[1]] = deflectorType
                deflatorsPlaced += 1
                
                // Update direction based on deflector
                currentDir = getNewDirection(currentDirection: currentDir, deflectorType: deflectorType)
                print("🔧 Placed deflector \(deflectorType) at [\(nextPos[0]),\(nextPos[1])], new direction: \(currentDir.name)")
            }
            
            currentPos = nextPos
        }
        
        // Place remaining deflectors randomly
        while deflatorsPlaced < deflectorCount {
            let row = Int.random(in: 0..<matrixSize)
            let col = Int.random(in: 0..<matrixSize)
            
            if matrix[row][col] == DeflectorTypes.empty {
                let deflectorType = Bool.random() ? DeflectorTypes.slash : DeflectorTypes.backslash
                matrix[row][col] = deflectorType
                deflatorsPlaced += 1
            }
        }
        
        return PlacementResult(success: true)
    }
    
    /**
     * Simulate ball path through matrix
     */
    private static func simulateBallPath(
        matrix: [[Int]],
        startPosition: [Int],
        startDirection: Direction,
        maxBounces: Int
    ) -> PathResult {
        let matrixSize = matrix.count
        var path: [BallPathStep] = []
        var activeDeflectors: [ActiveDeflector] = []
        
        var currentPos = startPosition
        var currentDir = startDirection.copy()
        var bounces = 0
        
        // Add start position
        path.append(BallPathStep(
            position: currentPos,
            direction: currentDir.copy(),
            deflected: false,
            deflectorType: nil
        ))
        
        while bounces < maxBounces {
            // Calculate next position
            let nextPos = [
                currentPos[0] + currentDir.dr,
                currentPos[1] + currentDir.dc
            ]
            
            // Check if out of bounds
            if nextPos[0] < 0 || nextPos[0] >= matrixSize ||
               nextPos[1] < 0 || nextPos[1] >= matrixSize {
                break
            }
            
            currentPos = nextPos
            let cellValue = matrix[currentPos[0]][currentPos[1]]
            
            // Check for deflector
            if cellValue == DeflectorTypes.slash || cellValue == DeflectorTypes.backslash {
                // Ball hits deflector
                let newDirection = getNewDirection(currentDirection: currentDir, deflectorType: cellValue)
                
                path.append(BallPathStep(
                    position: currentPos,
                    direction: currentDir.copy(),
                    deflected: true,
                    deflectorType: cellValue
                ))
                
                activeDeflectors.append(ActiveDeflector(
                    position: currentPos,
                    type: cellValue,
                    oldDirection: currentDir.copy(),
                    newDirection: newDirection.copy()
                ))
                
                currentDir = newDirection
                bounces += 1
            } else {
                // Empty cell
                path.append(BallPathStep(
                    position: currentPos,
                    direction: currentDir.copy(),
                    deflected: false,
                    deflectorType: nil
                ))
            }
        }
        
        let endPosition = path.last?.position ?? currentPos
        
        return PathResult(
            success: true,
            path: path,
            endPosition: endPosition,
            activeDeflectors: activeDeflectors
        )
    }
    
    /**
     * Calculate new direction after hitting deflector - MATCHES ANDROID EXACTLY
     */
    private static func getNewDirection(currentDirection: Direction, deflectorType: Int) -> Direction {
        switch deflectorType {
        case DeflectorTypes.slash: // Slash / deflector - EXACTLY like Android
            switch currentDirection.name {
            case "UP":
                return directions["RIGHT"]!
            case "RIGHT":
                return directions["UP"]!
            case "DOWN":
                return directions["LEFT"]!
            case "LEFT":
                return directions["DOWN"]!
            default:
                return getDirectionFromComponents(dr: -currentDirection.dc, dc: -currentDirection.dr)
            }
        case DeflectorTypes.backslash: // Backslash \ deflector - EXACTLY like Android
            switch currentDirection.name {
            case "UP":
                return directions["LEFT"]!
            case "LEFT":
                return directions["UP"]!
            case "DOWN":
                return directions["RIGHT"]!
            case "RIGHT":
                return directions["DOWN"]!
            default:
                return getDirectionFromComponents(dr: currentDirection.dc, dc: currentDirection.dr)
            }
        default:
            return currentDirection // No change
        }
    }
    
    /**
     * Get direction object from components
     */
    private static func getDirectionFromComponents(dr: Int, dc: Int) -> Direction {
        switch (dr, dc) {
        case (-1, 0):
            return directions["UP"]!
        case (1, 0):
            return directions["DOWN"]!
        case (0, -1):
            return directions["LEFT"]!
        case (0, 1):
            return directions["RIGHT"]!
        default:
            return Direction(dr: dr, dc: dc, name: "CUSTOM_\(dr)_\(dc)")
        }
    }
    
    /**
     * Count deflectors in matrix
     */
    private static func countDeflectors(matrix: [[Int]]) -> Int {
        return matrix.flatMap { $0 }.filter { cell in
            cell == DeflectorTypes.slash || cell == DeflectorTypes.backslash
        }.count
    }
    
    /**
     * Build JSON string from nested data structures
     */
    private static func buildJsonString(data: Any) -> String {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: data, options: [])
            return String(data: jsonData, encoding: .utf8) ?? "{}"
        } catch {
            print("Error building JSON: \(error)")
            return "{}"
        }
    }
    
    // Helper data structures
    struct PuzzleResult {
        let success: Bool
        let matrix: [[Int]]?
        let startPosition: [Int]?
        let startDirection: Direction?
        let ballPath: [BallPathStep]?
        let endPosition: [Int]?
        let activeDeflectors: [ActiveDeflector]?
        let error: String?
        
        init(success: Bool, matrix: [[Int]]? = nil, startPosition: [Int]? = nil,
             startDirection: Direction? = nil, ballPath: [BallPathStep]? = nil,
             endPosition: [Int]? = nil, activeDeflectors: [ActiveDeflector]? = nil,
             error: String? = nil) {
            self.success = success
            self.matrix = matrix
            self.startPosition = startPosition
            self.startDirection = startDirection
            self.ballPath = ballPath
            self.endPosition = endPosition
            self.activeDeflectors = activeDeflectors
            self.error = error
        }
    }
    
    struct StartPositionResult {
        let success: Bool
        let position: [Int]?
        let direction: Direction?
    }
    
    struct PlacementResult {
        let success: Bool
    }
    
    struct PathResult {
        let success: Bool
        let path: [BallPathStep]?
        let endPosition: [Int]?
        let activeDeflectors: [ActiveDeflector]?
    }
}
